# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [18]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime
%matplotlib inline
os.getcwd()

'/Users/ktregier/Desktop/Springboard/Unit 6/Capstone1/WeekendPrices'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [19]:
os.chdir("../WeekendPrices")
filepath = 'data/weekend3_output.csv'
df = pd.read_csv(filepath, header=0, index_col=0)
df.head()

Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0        0.0   
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
3     Arizona Snowbowl  Arizona        11500           2300      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  Snow Making_ac  \
0          0          2     2       0  ...             1610.0           113.0   
1          0          0     0       0  ...              640.0            60.0   
2          0          0     0       1  ...               30.0            30.0   
3          1          0     2       2  ...              777.0           104.0   
4          0          1     2       3  ...              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           550.0         1  
1               90.0             0.0         1  
2              152.0            30.0         1  
3              122.0             0.0         0  
4              104.0            80.0         0  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [20]:
#dfo = df.select_dtypes(include=['object'])
df = pd.concat([df.drop ('state', axis=1), pd.get_dummies(df.state)], axis=1)

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [21]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekday'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekday

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [22]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [23]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [24]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [8]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [25]:
from sklearn.metrics import mean_absolute_error, explained_variance_score
from math import sqrt
mae = mean_absolute_error(y_test, y_pred)
print("MAE :" , mae)

MAE : 5.202681836770903


In [26]:
evs = explained_variance_score( y_test, y_pred)
print("EVS :" ,evs)

EVS : 0.9353652288816366


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [27]:
print(lm.intercept_)

58.000097053741364


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [28]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
Coeff = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
Coeff.sort_values('Coefficient' , ascending = False)
#result.head(25)

Coefficient
total_chairs       3.453359e+13
fastQuads          1.309176e+13
surface            1.226600e+13
double             1.080925e+13
triple             9.642596e+12
New York           9.560562e+12
Michigan           9.022574e+12
Colorado           7.949411e+12
quad               7.814967e+12
California         7.779239e+12
Pennsylvania       7.423440e+12
New Hampshire      6.844994e+12
Wisconsin          6.844994e+12
Vermont            6.638182e+12
Minnesota          6.423264e+12
Utah               6.199398e+12
Montana            5.965576e+12
Idaho              5.965576e+12
Massachusetts      5.720578e+12
Washington         5.462900e+12
Oregon             5.462900e+12
Maine              5.190653e+12
New Mexico         5.190653e+12
Wyoming            4.901411e+12
North Carolina     4.257909e+12
Connecticut        3.892915e+12
Ohio               3.892915e+12
fastSixes          3.881059e+12
West Virginia      3.487282e+12
Virginia           3.487282e+12
Illinois           3.487282e+12
Nevada             3.487282e+12
trams              3.334714e+12
Alaska             3.024703e+12
Iowa               3.024703e+12
Missouri           2.473433e+12
New Jersey         2.473433e+12
South Dakota       2.473433e+12
Arizona            2.473433e+12
Indiana            2.473433e+12
Tennessee          1.751645e+12
Maryland           1.751645e+12
Rhode Island       1.751645e+12
fastEight          3.278350e+11
AdultWeekend       2.186865e+01
Runs               2.141022e+00
daysOpenLastYear   8.718262e-01
projectedDaysOpen  7.333984e-01
vertical_drop      7.154252e-01
Snow Making_ac     6.921921e-01
NightSkiing_ac     6.562500e-01
averageSnowfall    6.547852e-01
LongestRun_mi      4.916382e-01
yearsOpen          4.375000e-01
TerrainParks       3.115234e-01
SkiableTerrain_ac  1.787109e-01
clusters           9.057617e-02
summit_elev        1.835944e-02

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [29]:
model2 = pd.read_csv(filepath, header=0, index_col=0)
model2X = model2.drop(['AdultWeekday', 'Name', 'state'], axis=1)
model2y = model2.AdultWeekday
scaler = preprocessing.StandardScaler().fit(model2X)
X2_scaled=scaler.transform(model2X)
y2 = model2y.ravel()
X2_train, X2_test, y2_train, y2_test = train_test_split(X2_scaled, y2, test_size=0.25, random_state=1)
lm2 = linear_model.LinearRegression()
m2 = lm2.fit(X2_train, y2_train)
y2_pred = m2.predict(X2_test)
evs2 = explained_variance_score(y2_test, y2_pred)
mae2 = mean_absolute_error(y2_test, y2_pred)
print("ESV 2: ", evs2)
print("MAE: " , mae2 )
print("Intercept: ", lm2.intercept_)

ESV 2:  0.9170800948573538
MAE:  5.538589922672636
Intercept:  58.052002736857155


In [30]:
Coeff2 = pd.DataFrame(abs(lm2.coef_), model2X.columns, columns=['Coefficient'])
Coeff2.sort_values('Coefficient' , ascending = False)

Coefficient
AdultWeekend         21.177485
clusters              4.159832
summit_elev           2.767513
Snow Making_ac        1.816413
SkiableTerrain_ac     1.463993
averageSnowfall       1.337743
Runs                  1.312064
NightSkiing_ac        1.052505
fastSixes             0.986194
surface               0.952584
triple                0.894458
projectedDaysOpen     0.846486
daysOpenLastYear      0.816959
fastEight             0.665924
trams                 0.511924
fastQuads             0.509569
quad                  0.492623
yearsOpen             0.441104
total_chairs          0.422197
vertical_drop         0.322092
TerrainParks          0.292079
LongestRun_mi         0.254608
double                0.126335

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [31]:
model3 = pd.read_csv(filepath, header=0, index_col=0)
model3.columns
model3X = model3.drop(['AdultWeekday', 'Name', 'state', 'summit_elev'], axis=1)
model3y = model3.AdultWeekday
scaler = preprocessing.StandardScaler().fit(model3X)
X3_scaled=scaler.transform(model3X)
y3 = model3y.ravel()
X3_train, X3_test, y3_train, y3_test = train_test_split(X3_scaled, y3, test_size=0.25, random_state=1)
lm3 = linear_model.LinearRegression()
m3 = lm3.fit(X3_train, y3_train)
y3_pred = m3.predict(X3_test)
evs3 = explained_variance_score(y3_test, y3_pred)
mae3 = mean_absolute_error(y3_test, y3_pred)
print("ESV: ", evs3 )
print("MAE: ", mae3)
print("Intercept: ", lm3.intercept_)

ESV:  0.9185074205054775
MAE:  5.531549189044149
Intercept:  58.05325789390488


In [32]:
Coeff3 = pd.DataFrame(abs(lm3.coef_), model3X.columns, columns=['Coefficient'])
Coeff3.sort_values('Coefficient' , ascending = False)

Coefficient
AdultWeekend         20.900254
clusters              2.297560
Snow Making_ac        1.742646
averageSnowfall       1.262723
Runs                  1.210851
daysOpenLastYear      1.086510
projectedDaysOpen     1.033873
fastSixes             0.985611
triple                0.870871
SkiableTerrain_ac     0.836781
surface               0.816049
NightSkiing_ac        0.804383
fastEight             0.561571
quad                  0.542174
trams                 0.391818
yearsOpen             0.387398
total_chairs          0.372697
fastQuads             0.260283
vertical_drop         0.242749
LongestRun_mi         0.206413
TerrainParks          0.169843
double                0.140824

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.935365 | 5.202682 |-|
| Model 2. | 0.917080 | 5.538590 |'state'|
| Model 3. | 0.918507 | 5.531549 |'state','summit_elev','base_elev'|

In [33]:
model = pd.DataFrame({ "EVS": [evs, evs2, evs3], "MAE": [mae, mae2, mae3]}, index=[1, 2, 3])
model.head()

EVS       MAE
1  0.935365  5.202682
2  0.917080  5.538590
3  0.918507  5.531549

Model Selection:

The intercept value for all three models is similar.

I thought it was interesting that the model with the best performance (ss I understand it) is model 1. I guess it indicates that non-changable features (location and elevation) have an effect on ticket price.

I would probably choose model 3 for the basis of my presentation. It explains slightly more variation with slightly less error than model 2, and doesn't include the unchangeable factors in model 1.